src/kubernetes.cc

```diff
@@ -834,14 +848,26 @@ void KubernetesReader::WatchMaster(
           if (ec != boost::asio::error::operation_aborted) {
             watch_service->stop();
           }
+          });
+      }
+      std::unique_ptr<DelayTimer> silence_timer;
+      if (config_.KubernetesUpdaterWatchReconnectMaxSilenceSeconds() > 0) {
+        silence_timer = delay_timer_factory_->CreateTimer(*watch_service);
+        silence_timer->RunAsyncAfter(std::chrono::seconds(
+            config_.KubernetesUpdaterWatchReconnectMaxSilenceSeconds()),
+            [watch_service](boost::system::error_code const &ec) {
+          if (ec != boost::asio::error::operation_aborted) {
+            watch_service->stop();
+          }
         });
       }
       Watcher watcher(
         endpoint,
         [=](json::value raw_watch) {
           WatchEventCallback(callback, name, std::move(raw_watch));
         },
-        watch_service, verbose);
+        watch_service, timer.get(),
```

Comment:
1. I think this should be `silence_timer`, not `timer`, based on the variable name of watcher's Constructor.
2. maybe rename `timer` to other, so we won't have confusion between timer for interval reconnection & silence re-connection?  
3. It's bad smell to me that we mix `unique_ptr` with raw pointer (`timer.get()`) here, especially that the raw pointer will be used in async environment. Is it the lifecycle of `timer` well managed here? As the scope of `timer` is within the for-loop, if any error is thrown in this loop but the watcher doesn't know it yet, it may potentially cause crash. maybe consider using `shared_ptr` for this timer?

Reply:
3 is a very good point. We could, in fact, pass ownership of the timer to the watcher, since we don't reference it anywhere here and it would live as long as the watcher does anyway.

Also note that this is the wrong timer — we should pass `silence_timer` into the watcher, not `timer`.

Reply:
passing ownership of `silence_timer` sounds good to me!

Reply:
Done.